sycl/doc/extensions/proposed/sycl_ext_oneapi_device_image_content.asciidoc

```diff
@@ -0,0 +1,254 @@
+  format.
+
+:ref1: ../proposed/sycl_ext_oneapi_free_function_kernels.asciidoc#level-zero-and-opencl-compatibility
+
+[_Note:_ The interface to kernels in the device image is not defined in the
+general case, which means there is no portable way to invoke kernels from a
+Level Zero module that is created from the raw device image content.
+However, see link:{ref1}[here] for a limited case where this portability is
+guaranteed.
+_{endnote}_]
+
+=== Format on OpenCL
+
+The format depends on the `State` of the device image:
+
+* `bundle_state::input`: The format could change in the future, but it is
+  currently a SPIR-V module as described above for Level Zero.
+
+* `bundle_state::object`: The device image content is an unspecified format that
+  is created by calling `clCompileProgram`.
+
+* `bundle_state::executable`: The device image content is executable binary
+  device code representing one or more kernels, which can be passed to
+  `clCreateProgramWithBinary`.
+
+[_Note:_ The interface to kernels in the device image is not defined in the
+general case, which means there is no portable way to invoke kernels from a
+OpenCL `cl_program` object that is created from the raw device image content.
+However, see link:{ref1}[here] for a limited case where this portability is
+guaranteed.
+_{endnote}_]
+
+=== Format on CUDA
+
+The format depends on the `State` of the device image:
+
+* `bundle_state::input`: The device image content is a PTX module representing
+  one or more kernels.
+
+* `bundle_state::object`: ???
+
+* `bundle_state::executable`: The device image content is a CUBIN module
+  representing one or more kernels.
+
+
+== Example
+
+:ref2: ../proposed/sycl_ext_oneapi_free_function_kernels.asciidoc
+
+A kernel bundle can contain multiple device images with different
+representations of the same kernel for different devices.
+This example shows how to get the device image content for a particular kernel
+for a particular device.
+Note that this example also uses the kernel syntax described in link:{ref2}[
+sycl_ext_oneapi_free_function_kernels], but it is not necessary to define
+kernels in that syntax when using this extension.
+
+[source,c++]
+----
+#include <sycl/sycl.hpp>
+namespace syclext = sycl::ext::oneapi;
+namespace syclexp = sycl::ext::oneapi::experimental;
+
+SYCL_EXT_ONEAPI_FUNCTION_PROPERTY((syclexp::nd_range_kernel<1>))
+void iota(float start, float *ptr) {
+  size_t id = syclext::this_work_item::get_nd_item().get_global_linear_id();
+  ptr[id] = start + static_cast<float>(id);
+}
+
+void main() {
+  sycl::device d;
+  sycl::queue q{d};
+  sycl::context ctxt = q.get_context();
+
+  // Get a kernel bundle that contains the kernel "iota".
+  sycl::kernel_id iota = syclexp::get_kernel_id<iota>();
+  auto exe_bndl =
+    sycl::get_kernel_bundle<sycl::bundle_state::executable>(ctxt, {iota});
+
+  std::vector<std::byte> bytes;
+  for (auto& img: bundle) {
```

Comment:
Just for my education, are there any cases where there will be more than 1 image returned for this case? Except the multi-device case

Reply:
This is a good question!  For DPC++, this will be the case (at least for now).

@AerialMantis proposed the initial draft for the kernel bundle APIs, and he originally thought there could be multiple device images for a single device.  I think the idea was that the compiler could create multiple versions that are optimized for different scenarios.  I think this was just a hypothetical use case, though.  DPC++ does not currently do this.  If we did add this in the future, we would need some sort of query to distinguish between them.

Reply:
Oh wait.  Are you asking if a `kernel_bundle` could contain more than one device image when the system has just one device?  Yes, that can definitely happen.  Consider, for example, a case where you ask for a bundle that contains two kernels:

```
sycl::kernel_id foo = syclexp::get_kernel_id<foo>();
sycl::kernel_id bar = syclexp::get_kernel_id<bar>();
auto exe_bndl =
    sycl::get_kernel_bundle<sycl::bundle_state::executable>(ctxt, {foo, bar});
```

There is no guarantee that `foo` and `bar` are in the same device image, so the kernel bundle might contain two device images in this case.  The compiler decides how to pack kernels together into device images depending on the compilation flags (e.g. `-fsycl-device-code-split`) and also some internal algorithms.